packages/shylu/shylu_node/tacho/unit-test/Tacho_TestSolver.cpp

```diff
@@ -0,0 +1,242 @@
+// clang-format off
+// @HEADER
+// *****************************************************************************
+//                            Tacho package
+//
+// Copyright 2022 NTESS and the Tacho contributors.
+// SPDX-License-Identifier: BSD-2-Clause
+// *****************************************************************************
+// @HEADER
+// clang-format on
+#include <gtest/gtest.h>
+#include <Kokkos_Core.hpp>
+
+#include "Tacho_CommandLineParser.hpp"
+#include "Tacho_CrsMatrixBase.hpp"
+#include "Tacho_Driver.hpp"
+#include "Tacho_MatrixMarket.hpp"
+
+
+/// matrix generator
+template <typename CrsMatrixBaseTypeHost>
+int generate2dLaplace(int nx, CrsMatrixBaseTypeHost &A) {
+  // generate Laplace matrix on 5pt stencile
+  int n = nx*nx;
+  int nnz = n + 4*nx*(nx-1);
+
+  using value_type = typename CrsMatrixBaseTypeHost::value_type;
+  typename CrsMatrixBaseTypeHost::size_type_array    ap("ap", n+1);
+  typename CrsMatrixBaseTypeHost::ordinal_type_array aj("aj", nnz);
+  typename CrsMatrixBaseTypeHost::value_type_array   ax("ax", nnz);
+
+  nnz = 0;
+  ap(0) = 0;
+  for (int i=0; i<nx; i++) {
+    for (int j=0; j<nx; j++) {
+      int k = i*nx + j;
+      if (i > 0) {
+        aj(nnz) = k-nx;
+        ax(nnz) = value_type(1.0);
+        nnz ++;
+      }
+      if (j > 0) {
+        aj(nnz) = k-1;
+        ax(nnz) = value_type(1.0);
+        nnz ++;
+      }
+      aj(nnz) = k;
+      ax(nnz) = value_type(4.0);
+      nnz++;
+      if (j < nx-1) {
+        aj(nnz) = k+1;
+        ax(nnz) = value_type(1.0);
+        nnz ++;
+      }
+      if (i < nx-1) {
+        aj(nnz) = k+nx;
+        ax(nnz) = value_type(1.0);
+        nnz ++;
+      }
+      ap(k+1) = nnz;
+    }
+  }
+  A.clear();
+  A.setExternalMatrix(n, n, nnz, ap, aj, ax);
+  return nnz;
+}
+
+
+// main test driver
+template <typename value_type>
+int driver(std::string file, std::string method_name, int variant, int nrhs) {
```

Comment:
If you make any more changes on the PR, you could make each of the args `const` , don't worry if the PR is finalized though
```suggestion
int driver(const std::string file, const std::string method_name, const int variant, const int nrhs) {
```